### ISSUE TITLE
reportReturnType incorrectly merges types resulting in error

### ISSUE THREAD

**Issue by Gabgobie** [microsoft/pyright#10898]
## Environment data

-   Pylance version: `ms-python.vscode-pylance@2025.7.1`
-   OS and version: Windows 10 Pro 22H2
-   Python version (& distribution if applicable, e.g. Anaconda): `uv venv` with version `3.12.8`

## Code Snippet

```python
from typing import SupportsIndex


class Example:
    a: list[int]
    b: list[int]

    def __init__(self, a: list[int], b: list[int]):
        # assert len(a) == len(b)
        self.a = a
        self.b = b

    def __getitem__(self, val: SupportsIndex | slice) -> tuple[list[int], list[int]] | tuple[int, int]:
        return (self.a[val], self.b[val])  # Will merge the possible tuple results to an incorrect return type

```

Instead writing the following fixes the issue at the cost of unnecessary code and probably performance

```python
    def __getitem__(self, val: SupportsIndex | slice) -> tuple[list[int], list[int]] | tuple[int, int]:
        if isinstance(val, slice):  # This does nothing but is needed to fix the inferred return type
            return (self.a[val], self.b[val])
        else:
            return (self.a[val], self.b[val])
```

## Expected behavior

Don't incorrectly merge the possible resulting tuples to expect impossible permutations

## Actual behavior

In both cases, permutations such as `tuple[int, list[int]]` are not possible but pylance incorrectly infers a mixture in the first example.

## Logs

The logs contain personally identifiable information (Paths). The above should be sufficient to recreate the issue entirely.

**Comment by erictraut** [comment]
Pyright is working as intended here, so this isn't a bug.

A static type checker bases its analysis on the type information provided. It cannot safely assume that the expression `self.a[val]` and `self.b[val]` will return the same type. Not surprisingly, other type checkers like mypy report this same error.

**Comment by Gabgobie** [comment]
I get where you're coming from but don't quite agree.

The issue is not that `self.a` and `self.b` may yield different return types based on the same input but that it doesn't understand that `val` will be of the same type in both cases.

Therefore it evaluates `self.a[SupportsIndex | slice], self.b[SupportsIndex | slice]` instead of `self.a[SupportsIndex], self.b[SupportsIndex] | self.a[slice], self.b[slice]`.

Are you sure this ***should be*** the intended behavior?

**Comment by erictraut** [comment]
The type of `val` is the same in both cases: `SupportsIndex | slice`. When a static type checker sees the expression `self.a[val]`, it first evaluates the LHS type, which is `list[int]`. Then it evaluates `list[int].__getitem__` using the definition from the `list` class. This results in the type `int | list[int]`. The same evaluation is performed for the expression `self.b[val]` which also yields a type of `int | list[int]`. 

So yes, I'm sure this is the intended behavior. Any other behavior would violate the Python typing spec.

**Comment by Gabgobie** [comment]
Thanks for the explanation.

If you don't mind one more question: What is the correct way to deal with this? Overloading the method fixed errors in consumers but not with the original method. Should I just tolerate the `reportReturnType` using `# type: ignore`, take the hit from unnecessary branching or is there another way I'm not aware of?

**Comment by erictraut** [comment]
I'd typically use an overload in this case because provides finer-grained type information for callers.

```python
    @overload
    def __getitem__(self, val: SupportsIndex) -> tuple[int, int]: ...
    @overload
    def __getitem__(self, val: slice) -> tuple[list[int], list[int]]: ...
    def __getitem__(
        self, val: SupportsIndex | slice
    ) -> tuple[int | list[int], int | list[int]]:
        return (self.a[val], self.b[val])
```
